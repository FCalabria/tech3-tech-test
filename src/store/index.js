import Vue from 'vue'
import Vuex from 'vuex'

import transferHistory from './modules/transferHistory'
import authorization from './modules/authorization'

Vue.use(Vuex)

export default new Vuex.Store({
  modules: {
    transferHistory,
    authorization
  }
})
