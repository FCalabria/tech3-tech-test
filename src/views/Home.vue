<template>
  <div :class="`h-screen grid p-3 gap-3 ${$attrs.itemIndex ? 'md:grid-cols-2' : ''}`">
    <Spinner
      v-if="loading"
      class="justify-self-center place-self-center"
    />
    <ItemsList
      v-else
      :data="transferData"
    />
    <div v-if="selectedItem">
      <ItemDetail
        :debtor="selectedItem.attributes.debtor"
        :beneficiary="selectedItem.attributes.beneficiary"
      />
      <SimilarEntries />
    </div>
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex'
import ItemsList from '@/components/ItemsList.vue'
import ItemDetail from '@/components/ItemDetail.vue'
import SimilarEntries from '@/components/SimilarEntries.vue'
import Spinner from 'vue-simple-spinner'

export default {
  name: 'Home',
  components: {
    ItemsList,
    ItemDetail,
    SimilarEntries,
    Spinner
  },
  computed: {
    ...mapState({
      transferData: state => state.transferHistory.data,
      loading: state => state.transferHistory.loading
    }),
    selectedItem: function () {
      if (!this.transferData || !this.$attrs.itemIndex) {
        return {}
      } else {
        return this.transferData[this.$attrs.itemIndex]
      }
    }
  },
  created() {
    this.fetchTransferHistory()
  },
  methods: {
    ...mapActions({
      fetchTransferHistory: 'transferHistory/getAll'
    })
  }
}
</script>
