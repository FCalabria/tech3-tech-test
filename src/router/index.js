import Vue from 'vue'
import VueRouter from 'vue-router'
import Home from '../views/Home.vue'

Vue.use(VueRouter)

const routes = [
  {
    path: '/',
    name: 'Home',
    component: Home
  },
  {
    path: '/detail:itemId',
    name: 'Detail',
    props: true,
    // route level code-splitting
    // this generates a separate chunk (detail.[hash].js) for this route
    // which is lazy-loaded when the route is visited.
    component: Home
  }
]

const router = new VueRouter({
  routes
})

export default router
