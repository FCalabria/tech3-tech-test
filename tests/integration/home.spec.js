import fetch from 'just-fetch'
import VueCurrencyFilter from 'vue-currency-filter'
import { render, waitForElementToBeRemoved, fireEvent } from '@testing-library/vue'
import Home from '@/views/Home.vue'
import store from '@/store/index'
import router from '@/router/index'
import mockData from './mockData/getAll.json'

describe('Home', () => {
  fetch.get.mockReturnValue(Promise.resolve(mockData))
  let renderResult
  beforeEach(async () => {
    renderResult = render(Home, {
      store,
      routes: router,
      mocks: {
        $d: value => value,
        $t: value => value
      }
    }, vue => {
      vue.use(VueCurrencyFilter)
    })
    await waitForElementToBeRemoved(() => renderResult.queryByTestId('loading-spinner'))
  })
  test('it should show the list of transfers', () => {
    const result = renderResult.html()
    expect(result).toMatchSnapshot()
  })
  test('it should show the detail when a "view detail" link is clicked', async () => {
    const { getAllByRole, queryByTestId } = renderResult
    const links = getAllByRole('link', { name: 'actions.detail' })
    expect(queryByTestId('detail-section')).toBeNull()
    await fireEvent.click(links[2])
    const detail = queryByTestId('detail-section')
    expect(detail).not.toBeNull()
    expect(detail).toMatchSnapshot()
  })
})