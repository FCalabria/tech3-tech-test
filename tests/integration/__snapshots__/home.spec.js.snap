// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home it should show the detail when a "view detail" link is clicked 1`] = `
<div
  class="flex flex-col"
  data-testid="detail-section"
>
  <div
    class="md:space-y-4 flex md:block space-x-2 md:space-x-0 items-stretch"
  >
    <div
      class="shadow rounded-sm flex-grow"
    >
      <div
        class="p-3 flex items-center bg-blue-100"
      >
        <h2
          class="font-bold"
        >
          
      debtor
    
        </h2>
      </div>
       
      <div
        class="p-3"
      >
        <p>
          <span
            class="text-gray-500"
          >
            person.name:
          </span>
           Deborah Mcintosh
        </p>
         
        <p>
          <span
            class="text-gray-500"
          >
            person.sortCode:
          </span>
           75-23-24
        </p>
         
        <p>
          <span
            class="text-gray-500"
          >
            person.account:
          </span>
           72961387
        </p>
      </div>
    </div>
     
    <div
      class="shadow rounded-sm flex-grow"
    >
      <div
        class="p-3 flex items-center bg-blue-100"
      >
        <h2
          class="font-bold"
        >
          
      beneficiary
    
        </h2>
      </div>
       
      <div
        class="p-3"
      >
        <p>
          <span
            class="text-gray-500"
          >
            person.name:
          </span>
           Townsend Wood
        </p>
         
        <p>
          <span
            class="text-gray-500"
          >
            person.sortCode:
          </span>
           64-23-62
        </p>
         
        <p>
          <span
            class="text-gray-500"
          >
            person.account:
          </span>
           64008846
        </p>
      </div>
    </div>
  </div>
   
  <div
    class="shadow rounded-sm mt-4 overflow-hidden flex flex-col"
  >
    <div
      class="p-3 flex items-center bg-blue-100"
    >
      <h2
        class="font-bold"
      >
        
      similarEntries
    
      </h2>
    </div>
     
    <div
      class="p-3 overflow-auto"
    >
      <ul
        class="space-y-1"
      >
        <li>
          <a
            class=""
            href="#/detail/0"
          >
            
          968957cd-0cfb-4e7c-8009-0401b55e6fd9
        
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Home it should show the list of transfers 1`] = `
<div class="home h-screen grid md:grid-rows-1 p-3 gap-5 overflow-hidden grid-rows-1">
  <div class="items-list space-y-4 overflow-y-auto">
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">0</span> <span class="font-bold mx-3 capitalize">ADMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Mon Nov 24 2003 11:39:50 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/0" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 3.212</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">1</span> <span class="font-bold mx-3 capitalize">SUBMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Mon Jun 04 2018 23:52:31 GMT+0200 (Central European Summer Time)</time> <a href="#/detail/1" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 1.335</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">2</span> <span class="font-bold mx-3 capitalize">ADMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Sat May 08 2004 06:38:27 GMT+0200 (Central European Summer Time)</time> <a href="#/detail/2" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 9.044</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">3</span> <span class="font-bold mx-3 capitalize">SUBMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Sat Jun 24 2006 05:30:05 GMT+0200 (Central European Summer Time)</time> <a href="#/detail/3" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 1.939</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">4</span> <span class="font-bold mx-3 capitalize">REVERSAL</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Fri Dec 07 1990 22:44:19 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/4" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-green-800">ACCEPTED</span> <span class="text-lg"> 4.836</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">5</span> <span class="font-bold mx-3 capitalize">REVERSAL</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Tue Mar 04 1980 06:41:37 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/5" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 8.915</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">6</span> <span class="font-bold mx-3 capitalize">REVERSAL</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Thu Aug 28 2003 20:17:15 GMT+0200 (Central European Summer Time)</time> <a href="#/detail/6" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 1.512</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">7</span> <span class="font-bold mx-3 capitalize">ADMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Tue Nov 06 1984 14:05:59 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/7" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-green-800">ACCEPTED</span> <span class="text-lg"> 7.873</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">8</span> <span class="font-bold mx-3 capitalize">REVERSAL</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Mon Feb 08 1993 19:12:23 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/8" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-red-800">DECLINED</span> <span class="text-lg"> 5.398</span></div>
    </div>
    <div class="shadow rounded-sm item border-2 border-blue-900">
      <div class="p-3 flex items-center bg-blue-900 text-white"><span class="text-xs">9</span> <span class="font-bold mx-3 capitalize">ADMISSION</span>
        <div class="flex items-end md:items-center flex-col md:flex-row ml-auto"><time class="text-sm text-blue-200 order-2 md:order-1">Thu Dec 06 1979 11:00:24 GMT+0100 (Central European Standard Time)</time> <a href="#/detail/9" class="ml-3 link-negative order-1 md:order-2">
            actions.detail
          </a></div>
      </div>
      <div class="p-3 flex justify-between"><span class="text-green-800">ACCEPTED</span> <span class="text-lg"> 920</span></div>
    </div>
  </div>
  <!---->
</div>
`;
